core/preconditioner/isai.cpp

```diff
@@ -127,38 +170,94 @@ void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
     // system of excess blocks
     Array<IndexType> excess_row_ptrs_full{exec, num_rows + 1};
 
-    exec->run(isai::make_generate_tri_inverse(
-        lend(to_invert), lend(inverted), excess_block_ptrs.get_data(),
-        excess_row_ptrs_full.get_data(), is_lower));
+    if (is_general || is_spd) {
+        exec->run(isai::make_generate_general_inverse(
+            lend(to_invert), lend(inverted), excess_block_ptrs.get_data(),
+            excess_row_ptrs_full.get_data(), is_spd));
+    } else {
+        exec->run(isai::make_generate_tri_inverse(
+            lend(to_invert), lend(inverted), excess_block_ptrs.get_data(),
+            excess_row_ptrs_full.get_data(), is_lower));
+    }
 
-    auto excess_dim =
+    auto total_excess_dim =
         exec->copy_val_to_host(excess_block_ptrs.get_const_data() + num_rows);
+    auto excess_lim = excess_limit == 0 ? total_excess_dim : excess_limit;
     // if we had long rows:
-    if (excess_dim > 0) {
-        // build the excess sparse triangular system
-        auto excess_nnz = exec->copy_val_to_host(
-            excess_row_ptrs_full.get_const_data() + num_rows);
-        auto excess_system =
-            Csr::create(exec, dim<2>(excess_dim, excess_dim), excess_nnz);
-        auto excess_rhs = Dense::create(exec, dim<2>(excess_dim, 1));
-        auto excess_solution = Dense::create(exec, dim<2>(excess_dim, 1));
-        exec->run(isai::make_generate_excess_system(
-            lend(to_invert), lend(inverted), excess_block_ptrs.get_const_data(),
-            excess_row_ptrs_full.get_const_data(), lend(excess_system),
-            lend(excess_rhs)));
-        // solve it after transposing
-        std::unique_ptr<LinOpFactory> trs_factory;
-        if (is_lower) {
-            trs_factory = UpperTrs::build().on(exec);
-        } else {
-            trs_factory = LowerTrs::build().on(exec);
+    if (total_excess_dim > 0) {
+        bool done = false;
+        size_type block = 0;
+        while (true) {
```

Comment:
can we write this in terms of a normal `while` loop instead? it seems the condition you are checking here is `while(block < num_rows)`